libtransmission/file-posix.c

```diff
@@ -430,6 +451,132 @@ bool tr_sys_path_rename(char const* src_path, char const* dst_path, tr_error** e
     return ret;
 }
 
+/* We try to do a fast (in-kernel) copy using a variety of non-portable system
+ * calls. If the current implementation does not support in-kernel copying, we
+ * use a user-space fallback instead. */
+bool tr_sys_path_copy(char const* src_path, char const* dst_path, uint64_t file_size, tr_error** error)
+{
+    TR_ASSERT(src_path != NULL);
+    TR_ASSERT(dst_path != NULL);
+
+#if defined(__APPLE__) && defined(HAVE_COPYFILE)
+    (void)file_size; /* unused variable */
+
+    copyfile_flags_t const flags =
+#ifdef COPYFILE_CLONE /* macos >= 10.12 */
+        COPYFILE_CLONE;
+#else
+        COPYFILE_DATA;
+#endif
+    if (copyfile(src_path, dst_path, NULL, flags) < 0)
+    {
+        set_system_error(error, errno);
+        return false;
+    }
+    else
+    {
+        return true;
+    }
+
+#else /* __APPLE__ && HAVE_COPYFILE */
+
+    /* Other OSes require us to copy between file descriptors, so open them. */
+
+    tr_sys_file_t in;
+    tr_sys_file_t out;
+
+    in = tr_sys_file_open(src_path, TR_SYS_FILE_READ |
+        TR_SYS_FILE_SEQUENTIAL, 0, error);
+
+    if (in == TR_BAD_SYS_FILE)
+    {
+        tr_error_prefix(error, "Unable to open old file: ");
+        return false;
+    }
+
+    out = tr_sys_file_open(dst_path, TR_SYS_FILE_WRITE | TR_SYS_FILE_CREATE |
+        TR_SYS_FILE_TRUNCATE, 0666, error);
+
+    if (out == TR_BAD_SYS_FILE)
+    {
+        tr_error_prefix(error, "Unable to open new file: ");
+        tr_sys_file_close(in, NULL);
+        return false;
+    }
+
+#if defined(USE_COPY_FILE_RANGE) || defined(USE_SENDFILE64)
+
+    while (file_size > 0)
+    {
+        size_t const chunk_size = MIN(file_size, SIZE_MAX);
```

Comment:
Using chunk size of `SIZE_MAX` or any value greater than `SSIZE_MAX` will lead to issues. Even if the API call below will succeed, it will be forced to return a negative value (given it's return type is `ssize_t` which is usually a signed equivalend of `size_t`).

Reply:
I don't think any of the kernel APIs do that; they will only return negative values on failure, so effectively they are capped at SSIZE_MAX. Watching in the debugger, even on x64 transfers are done in 2 GB chunks even for chunk_size > 2 GB. But there's no harm in choosing SSIZE_MAX.